lib/services/database/challenge_repository_service.dart

```diff
@@ -0,0 +1,222 @@
+    });
+  }
+
+  /// Returns the active challenges of the user with id [uid] who is located at [pos].
+  /// If challenges are expired or missing, they are replaced by new challenges in the
+  /// database, and then returned. This therefore always ensure this method will return
+  /// challenges, all up-to-date; trying its best to find [maxActiveChallenges] (if
+  /// enough exist close enough).
+  /// Warning: This method is NOT atomic, and therefore not thread-safe. We suppose here
+  /// no malicious user will try to load the challenge page on two different devices at
+  /// the same time. In fact, even if this happens, this should not be a big problem.
+  Future<List<ChallengeFirestore>> getChallenges(
+    UserIdFirestore uid,
+    GeoPoint pos,
+  ) async {
+    final userDocRef =
+        _firestore.collection(UserFirestore.collectionName).doc(uid.value);
+
+    return _getChallenges(userDocRef, pos);
+  }
+
+  Future<List<ChallengeFirestore>> _getChallenges(
+    DocumentReference parentRef,
+    GeoPoint pos,
+  ) async {
+    final Set<PostIdFirestore> justExpired = {};
+    final List<ChallengeFirestore> activeChallenges =
+        List.empty(growable: true);
+
+    await _removeOldChallenges(parentRef, activeChallenges, justExpired);
+
+    if (activeChallenges.length < maxActiveChallenges) {
+      await _generateNewChallenges(
+        activeChallenges,
+        pos,
+        justExpired,
+        parentRef,
+      );
+    }
+
+    return activeChallenges;
+  }
+
+  /// Adds active challenges to [activeChallenges]. Removes expired challenges
+  /// and adds them to [justExpired]
+  Future<void> _removeOldChallenges(
+    DocumentReference parentRef,
+    List<ChallengeFirestore> activeChallenges,
+    Set<PostIdFirestore> justExpired,
+  ) async {
+    final challengesCollectionRef = _activeChallengesRef(parentRef);
+    final challengesSnap = await challengesCollectionRef.get();
+
+    for (final challengeSnap in challengesSnap.docs) {
+      final challenge = ChallengeFirestore.fromDb(challengeSnap);
+      if (challenge.data.isExpired ||
+          !await _postRepositoryService.postExists(challenge.postId)) {
+        await _moveToPastChallenge(challengeSnap, parentRef);
+        justExpired.add(challenge.postId);
+      } else {
+        activeChallenges.add(challenge);
+      }
+    }
+  }
+
+  /// Generates new challenges and puts them in [activeChallenges]
+  Future<void> _generateNewChallenges(
+    List<ChallengeFirestore> activeChallenges,
+    GeoPoint pos,
+    Set<PostIdFirestore> justExpired,
+    DocumentReference parentRef,
+  ) async {
+    final now = DateTime.now();
+    final sum = now.add(maxChallengeDuration);
+    final expiresOn = DateTime(
+      sum.year,
+      sum.month,
+      sum.day,
+    ); // truncates to the day
+
+    final Iterable<PostIdFirestore> possiblePosts =
+        (await _inRangeUnsortedPosts(pos))
+            .where((post) => !justExpired.contains(post));
```

Comment:
Is this line really needed, all `justExpired` challenges are stored in the `pastChallenges`, aren't they? I'm not sure I understand the purpose of `justExpired` more generally

Reply:
You're right, it is a relic of a time when I was trying to read both collections before writing to them, because I was trying to use transactions. I removed `justExpired` completely. 57a97123fb4eda47c02905c1753477766ca7cbdc